### PR TITLE
[CSM] Add function to get the definition of nodes and items

### PR DESCRIPTION
**Pull request by bigfoot547**
* [x] Get item definition
* [x] Get node definition
* [x] Document

### CODE REVIEW

**Comment by bigfoot547** [comment]
The reason I have not finished the item definition api is because I can't return nested tables. Example:
```lua
{
    a = 1,
    b = 2,
    c = {
        c = 3
    },
    d = 4
}
```

**Comment by nerzhul** [comment]
@bigfoot547 i think there are functions for deserialize it, look at register_item core side function no ?

**Comment by bigfoot547** [comment]
Found it! Thanks @nerzhul!
https://github.com/minetest/minetest/blob/dc5bc6cac7b81ab27e0064bc25b5fd1d8d617340/src/script/common/c_content.h#L91

**Comment by nerzhul** [comment]
@bigfoot547 np :)

**Comment by bigfoot547** [comment]
I am *almost* done with item definitions! I just need to iron out a segfault.

**Comment by bigfoot547** [comment]
Item definitions now complete!
Now working on node definition.

**Comment by bigfoot547** [comment]
Finished nodedef. Now to write documentation!
Please review code @nerzhul.

**Comment by bigfoot547** [comment]
I completed the documentation! This is ready for merge.

File: doc/client_lua_api.md [hunk 1]
```diff
@@ -955,6 +955,12 @@ Can be obtained via `minetest.get_meta(pos)`.
     * `inventory`: `{list1 = {}, ...}}`
 
 -----------------
+### Definitions
+* `minetest.get_node_def(itemstring)`
```
Comment (nerzhul) [inline doc/client_lua_api.md]: itemstring => nodename

File: src/script/common/c_content.cpp [hunk 2]
```diff
@@ -128,6 +128,77 @@ void push_item_definition(lua_State *L, const ItemDefinition &i)
...
+void push_item_definition_full(lua_State *L, const ItemDefinition &i)
+{
+	std::string type;
+	// This code is pure garbage, but that is what reviewers are for!
+	const bool use_caps = i.type == ITEM_TOOL;
+
+	switch (i.type)
+	{
+		case ITEM_NONE:
+		{
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: all those braces are useless

File: src/script/common/c_content.cpp [hunk 3]
```diff
@@ -128,6 +128,77 @@ void push_item_definition(lua_State *L, const ItemDefinition &i)
...
+		case ITEM_CRAFT:
+		{
+			type = "craftitem";
+			break;
+		}
+		case ITEM_TOOL:
+		{
+			type = "tool";
+			break;
+		}
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: you don't handle the default case this is a problem. you should trigger a FATAL_ERROR

File: src/script/common/c_content.cpp [hunk 4]
```diff
@@ -669,6 +740,400 @@ ContentFeatures read_content_features(lua_State *L, int index)
...
+	std::string paramtype = "none";
+	std::string paramtype2 = "none";
+	std::string drawtype = "normal";
+	std::string liquid_type = "none";
+	
+	// paramtype
+	switch (c.param_type)
+	{
+		case CPT_NONE:
+		{
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: brace removal + default case

File: src/script/common/c_content.cpp [hunk 5]
```diff
@@ -669,6 +740,400 @@ ContentFeatures read_content_features(lua_State *L, int index)
...
+void push_content_features(lua_State *L, const ContentFeatures &c)
+{
+	std::string paramtype = "none";
+	std::string paramtype2 = "none";
+	std::string drawtype = "normal";
+	std::string liquid_type = "none";
+	
+	// paramtype
+	switch (c.param_type)
+	{
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: brace should be upper

File: src/script/lua_api/l_client.cpp [hunk 6]
```diff
@@ -251,6 +252,50 @@ int ModApiClient::l_get_server_info(lua_State *L)
...
+	IItemDefManager *idef = gdef->idef();
+	assert(idef);
+	
+	if (!lua_isstring(L, 1))
+		return 0;
+
+	const std::string &name = lua_tostring(L, 1);
+	if (!idef->isKnown(name))
+		return 0;
+	const ItemDefinition def = idef->get(name);
```
Comment (nerzhul) [inline src/script/lua_api/l_client.cpp]: it MUST be a const ref assignment

File: src/script/common/c_content.cpp [hunk 7]
```diff
@@ -669,6 +740,400 @@ ContentFeatures read_content_features(lua_State *L, int index)
...
+	
+	// paramtype2
+	switch (c.param_type_2)
+	{
+		case CPT2_NONE:
+		{
+			// paramtype2 already equals "none"
+			break;
+		}
+		case CPT2_FULL:
```
Comment (SmallJoker) [inline src/script/common/c_content.cpp]: Please stop the code duplication at this point and reuse the lookup table `ScriptApiNode::es_ContentParamType2` by moving to `nodedef`. (Same for the drawtypes)

File: src/script/common/c_content.cpp [hunk 8]
```diff
@@ -128,6 +128,77 @@ void push_item_definition(lua_State *L, const ItemDefinition &i)
 	lua_setfield(L, -2, "description");
 }
 
+void push_item_definition_full(lua_State *L, const ItemDefinition &i)
+{
+	std::string type;
+	// This code is pure garbage, but that is what reviewers are for!
```
Comment (SmallJoker) [inline src/script/common/c_content.cpp]: We're not just here to tell you that you can't write good code. Look up the code style guidelines or some newer created code parts to see how it should be done.

File: src/script/common/c_content.cpp (hunk 2 above)
Comment (bigfoot547) [inline src/script/common/c_content.cpp]: Does switch...case automatically add `break`?

Comment (Zeno-) [inline src/script/common/c_content.cpp]: No

**Comment by bigfoot547** [comment]
Addressed reviews! Please review again @nerzhul @SmallJoker! Thanks!

File: src/script/common/c_content.cpp [hunk 9]
```diff
@@ -128,6 +129,54 @@ void push_item_definition(lua_State *L, const ItemDefinition &i)
 	lua_setfield(L, -2, "description");
 }
 
+void push_item_definition_full(lua_State *L, const ItemDefinition &i)
+{
+	std::string type;
+	const bool use_caps = i.type == ITEM_TOOL;
+	
+	type = (std::string)es_ItemType[(int)i.type].str;
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: you can directly do this assignment in variable construction

File: src/script/common/c_content.cpp [hunk 10]
```diff
@@ -669,6 +718,222 @@ ContentFeatures read_content_features(lua_State *L, int index)
...
 
+void push_content_features(lua_State *L, const ContentFeatures &c)
+{
+	std::string paramtype = "none";
+	std::string paramtype2 = "none";
+	std::string drawtype = "normal";
+	std::string liquid_type = "none";
+	
+	// paramtype
+	paramtype = (std::string)ScriptApiNode::es_ContentParamType[(int)c.param_type].str;
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: please don't cast and assign directly at variable construction

File: src/script/common/c_content.cpp [hunk 11]
```diff
@@ -669,6 +718,222 @@ ContentFeatures read_content_features(lua_State *L, int index)
...
+}
+
+/******************************************************************************/
+void push_nodebox(lua_State *L, const NodeBox &box)
+{
+	lua_newtable(L);
+	switch (box.type)
+	{
+		case NODEBOX_REGULAR:
+		{
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: all braces can be removed there as you don't create variables in cases

File: src/script/lua_api/l_client.cpp [hunk 12]
```diff
@@ -251,6 +252,50 @@ int ModApiClient::l_get_server_info(lua_State *L)
...
+	IGameDef *gdef = getGameDef(L);
+	assert(gdef);
+
+	IItemDefManager *idef = gdef->idef();
+	assert(idef);
+	
+	if (!lua_isstring(L, 1))
+		return 0;
+
+	const std::string &name = lua_tostring(L, 1);
```
Comment (nerzhul) [inline src/script/lua_api/l_client.cpp]: i don't think this works on every compiler as lua_tostring provides a const char * and you have a const ref std::string, here a copy is required

File: src/script/common/c_content.cpp (hunk 10 above)
Comment (bigfoot547) [inline src/script/common/c_content.cpp]: What should I do instead?

**Comment by bigfoot547** [comment]
Please review again! Thanks!

**Comment by bigfoot547** [comment]
@nerzhul @SmallJoker

File: src/script/common/c_content.cpp [hunk 13]
```diff
@@ -128,6 +129,52 @@ void push_item_definition(lua_State *L, const ItemDefinition &i)
 	lua_setfield(L, -2, "description");
 }
 
+void push_item_definition_full(lua_State *L, const ItemDefinition &i)
+{
+	std::string type  = (std::string)es_ItemType[(int)i.type].str;
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: std::string type(es_ItemType[(int)i.type].str)

File: src/script/common/c_content.cpp (hunk 10 above)
Comment (nerzhul) [inline src/script/common/c_content.cpp]: es_ItemType[(int)i.type].str

Comment (nerzhul) [inline src/script/common/c_content.cpp]: std::string type(es_ItemType[(int)i.type].str)

File: src/script/common/c_content.cpp [hunk 14]
```diff
@@ -669,6 +716,214 @@ ContentFeatures read_content_features(lua_State *L, int index)
...
+			lua_setfield(L, -2, "connect_bottom");
+			push_box(L, box.connect_front);
+			lua_setfield(L, -2, "connect_front");
+			push_box(L, box.connect_back);
+			lua_setfield(L, -2, "connect_back");
+			push_box(L, box.connect_left);
+			lua_setfield(L, -2, "connect_left");
+			push_box(L, box.connect_right);
+			lua_setfield(L, -2, "connect_right");
+			break;
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: can you add  a default with FATAL_ERROR("Unhandled box.type")

**Comment by bigfoot547** [comment]
@nerzhul @SmallJoker

File: src/script/common/c_content.cpp [hunk 15]
```diff
@@ -128,6 +129,52 @@ void push_item_definition(lua_State *L, const ItemDefinition &i)
...
+	lua_setfield(L, -2, "color");
+	push_v3f(L, i.wield_scale);
+	lua_setfield(L, -2, "wield_scale");
+	lua_pushinteger(L, i.stack_max);
+	lua_setfield(L, -2, "stack_max");
+	lua_pushboolean(L, i.usable);
+	lua_setfield(L, -2, "usable");
+	lua_pushboolean(L, i.liquids_pointable);
+	lua_setfield(L, -2, "liquids_pointable");
+	if (use_caps) {
```
Comment (SmallJoker) [inline src/script/common/c_content.cpp]: `if (i.type == ITEM_TOOL) {`

File: src/script/common/c_content.cpp [hunk 16]
```diff
@@ -669,6 +716,205 @@ ContentFeatures read_content_features(lua_State *L, int index)
...
+	lua_pushboolean(L, c.climbable);
+	lua_setfield(L, -2, "climbable");
+	lua_pushboolean(L, c.buildable_to);
+	lua_setfield(L, -2, "buildable_to");
+	lua_pushboolean(L, c.rightclickable);
+	lua_setfield(L, -2, "rightclickable");
+	lua_pushnumber(L, c.damage_per_second);
+	lua_setfield(L, -2, "damage_per_second");
+	if (c.isLiquid())
+	{
```
Comment (SmallJoker) [inline src/script/common/c_content.cpp]: Bracket: `if (c.isLiquid()) {`

File: src/script/common/c_content.cpp [hunk 17]
```diff
@@ -669,6 +716,205 @@ ContentFeatures read_content_features(lua_State *L, int index)
...
+	}
+}
+
+void push_box(lua_State *L, const std::vector<aabb3f> &box)
+{
+	lua_newtable(L);
+	u8 i = 1;
+	for (std::vector<aabb3f>::const_iterator it = box.begin();
+		it != box.end(); ++it) {
+			push_aabb3f(L, (*it));
```
Comment (SmallJoker) [inline src/script/common/c_content.cpp]: Indents the other way:
```
	it != box.end(); ++it) {
push_aabb3f(L, *it);
```

**Comment by bigfoot547** [comment]
I'm sorry I'm being a pest, *buuuuut*, the feature freeze is in 2 days.
@nerzhul @SmallJoker

File: doc/client_lua_api.md [hunk 18]
```diff
@@ -955,6 +955,12 @@ Can be obtained via `minetest.get_meta(pos)`.
     * `inventory`: `{list1 = {}, ...}}`
 
 -----------------
+### Definitions
+* `minetest.get_node_def(nodename)`
+    * Returns node definition table of `nodename`
+* `minetest.get_item_def(itemstring)`
+    * Returns item definition table of `itemstring`
```
Comment (SmallJoker) [inline doc/client_lua_api.md]: Do we perhaps need a documentation for these definition tables in this file too?

**Comment by SmallJoker** [comment]
Did a shallow test. The functions give outputs as expected.

**Comment by bigfoot547** [comment]
@SmallJoker

**Comment by bigfoot547** [comment]
Yay! Thanks!

**Pull request merged**